### PR TITLE
refactor: replace Logger.LogMessage by storing an io.Writer

### PR DESCRIPTION
**Pull request by klaernie**
This replaces the bad design I made in #360 regarding the output of formerly error.PrintErrors, which became error.FormatErrors and is now again error.PrintErrors.

Passing around slices of `LogMessage` what not the best design - it allowed testing the new code, but the resulting code was unreadable.

Now we have a `Logger.writer` that automatically gets initialized to `os.Stdout`, but can be changed using `Logger.SetWriter()`. This allows us to restore the previous form of calling `Logger.Error` and friends directly, and still capture their output to make the tests work.

Along the way I noticed a few inconsistencies that needed to be fixed, I hope this PR is still reviewable.

### CODE REVIEW

File: pkg/logger/logger.go [hunk 1]
```diff
@@ -13,115 +15,103 @@ const (
...
 
-type LogMessage struct {
-	Level   string
-	Message string
-}
-
 // Logger struct
 type Logger struct {
 	Verbosee bool
 	Debugg   bool
```
Comment (ccoVeille) [inline pkg/logger/logger.go:21]: I know it's out of the current PR scope

But please consider adding a comment if it's "normal" to have Verbosee and Debugg

I'm sure it's possible to find a better suffix. Here it looks like a typo

File: pkg/logger/logger.go [hunk 2]
```diff
@@ -13,115 +15,103 @@ const (
...
-	fmt.Printf("%s", message)
-}
-
-// Println prints a message with a trailing newline
-func Println(message string) {
-	fmt.Printf("%s\n", message)
+// println prints a message with a trailing newline
+func (l Logger) println(message string) {
+	l.LazyInit()
+	fmt.Fprintf(l.writer, "%s\n", message)
```
Comment (ccoVeille) [inline pkg/logger/logger.go:104]:
```suggestion
	fmt.Fprintln(l.writer, message)
```

File: pkg/logger/logger.go [hunk 3]
```diff
@@ -13,115 +15,103 @@ const (
...
 }
 
-// PrintColor prints a message in a given ANSI-color
-func PrintColor(message string, color string) {
-	fmt.Printf("%s%s%s", color, message, RESET)
+// printColor prints a message in a given ANSI-color
+func (l Logger) printColor(message string, color string) {
+	l.LazyInit()
+	fmt.Fprintf(l.writer, "%s%s%s", color, message, RESET)
```
Comment (ccoVeille) [inline pkg/logger/logger.go:110]: RESET variable looks abnormal (its case)

First a full uppercase is wrongly discouraged in Go

Then it also means it's exported.

Do we really need this?

But this remark is out of PR scope as RED, YELLOW, RESET … were present before this current PR

File: pkg/logger/logger.go [hunk 4]
```diff
@@ -13,115 +15,103 @@ const (
...
+	return logger
+}
+
+// initialize the Logger to write to standard output
+func (l *Logger) Init() {
+	l.writer = os.Stdout
+}
+
+// ensure the Logger is initialized on first print
+func (l *Logger) LazyInit() {
```
Comment (ccoVeille) [inline pkg/logger/logger.go:38]:
```suggestion
func (l *Logger) lazyInit() {
```

I don't think you need it to be exported

File: pkg/logger/logger.go (hunk 1 above)
Comment (klaernie) [reply pkg/logger/logger.go]: I renamed them to `VerboseEnabled` and `DebugEnabled` - that seemed like a better name and avoiding that typo-esque suffix

File: pkg/logger/logger.go (hunk 2 above)
Comment (klaernie) [reply pkg/logger/logger.go]: good point. done.

File: pkg/logger/logger.go (hunk 3 above)
Comment (klaernie) [reply pkg/logger/logger.go]: I renamed them by normalizing their names and prefixing them with `escSeq` which should help people unfamiliar with escape sequences to recognize them as such.

File: pkg/logger/logger.go (hunk 4 above)
Comment (klaernie) [reply pkg/logger/logger.go]: You're totally right. fixed.

File: Makefile [hunk 5]
```diff
@@ -42,7 +42,7 @@ uninstall: ## Remove executable from PATH
 	rm -f $(DESTDIR)$(prefix)$(mandir)/man1/editorconfig-checker.1
 
 test: ## Run test suite
-	@go test -race -coverprofile=coverage.txt -covermode=atomic ./...
+	@go test -ldflags "-X github.com/editorconfig-checker/editorconfig-checker/v3/cmd/editorconfig-checker.version=$(CURRENT_VERSION)" -race -coverprofile=coverage.txt -covermode=atomic ./...
```
Comment (theoludwig) [inline Makefile:45]: Could you explain why this change is needed?

File: cmd/editorconfig-checker/main.go [hunk 6]
```diff
@@ -2,56 +2,69 @@
...
 	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/config"
 	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/error"
 	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/files"
-	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/logger"
 	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/outputformat"
 	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/utils"
 	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/validation"
 )
 
 // version is used vor the help
```
Comment (theoludwig) [inline cmd/editorconfig-checker/main.go:19]: Not related to this PR, but now that I see the typo, and that the PR is related to refactoring and code clean up anyway, might be worth to fix it:

```suggestion
// version is used for the help
```

File: cmd/editorconfig-checker/main_test.go
Comment (theoludwig) [inline cmd/editorconfig-checker/main_test.go:1]: Good to add tests for the main, and CLI options. As explained in https://github.com/editorconfig-checker/editorconfig-checker/issues/82#issuecomment-2361817709, it would be even better to test it agnostic to the framework or even more agnostic, to the programming languages used.

File: cmd/editorconfig-checker/main.go (hunk 6 above)
Comment (klaernie) [reply cmd/editorconfig-checker/main.go]: lol.. I know - once one sees it, one cannot unsee it.. I fixed it, and added a little to the explanation on what that variable is for.

File: Makefile (hunk 5 above)
Comment (klaernie) [reply Makefile:45]: This is one of the changes I had to make due to the issue I described in #383 - since with that PR we are starting to run the binary (or at least it's `main` function) we also tend to parse config files - which means we need to have the current version available for comparison - hence the injection of the overwriting the initializing value of that variable.

File: cmd/editorconfig-checker/main_test.go
Comment (klaernie) [reply cmd/editorconfig-checker/main_test.go:1]: I think this currently is as far as possible testing from outside the main function - yes, we hook into the `main` function by overriding it's `exitStub`, but in every other perspective we run `main` as a goroutine as independent of the test framework as possible - providing only `os.Args` and external files - calling main and collecting only the return code. I did not yet attempt to override `os.Stdout` for capturing end-to-end snapshot tests (that might become its own PR later on).

File: Makefile (hunk 5 above)
Comment (theoludwig) [reply Makefile:45]: Alright, thank you for the explanation. :+1:

File: cmd/editorconfig-checker/main_test.go
Comment (theoludwig) [reply cmd/editorconfig-checker/main_test.go:1]: Yes, let's keep the tests in Go for the moment, easier to track test coverage this way also.

**Review by theoludwig — APPROVED** [review]

**Comment by klaernie** [comment]
I should probably rebase this branch, since there were some updates to main. But I wonder - is this PR to big, and I should try to break it into a few PRs? Sadly I think I coded myself into a corner with that big of a change set..

**Comment by theoludwig** [comment]
> I should probably rebase this branch, since there were some updates to main. But I wonder - is this PR to big, and I should try to break it into a few PRs? Sadly I think I coded myself into a corner with that big of a change set..

Yeah, the PR is quite big, and quite hard to review with all the changes. But I think, it's fine now, I think before doing anything, we need to wait other reviewers comments.
Or as Max said, 2 reviewers needed for a PR might be "too much" (https://github.com/editorconfig-checker/editorconfig-checker/pull/378#pullrequestreview-2297480255), we might reduce it to 1.
This PR is already opened since 1 month, we might still wait 1 week, and if next week, there are still no activity from others people, I will open the PR for reducing the amount of reviewer needed for a PR, and then we can merge this one.
Do you think it's a good way to move forward?

Also, do you think, we should wait #385 to release v3.1.0? I feel like there are already lot of changes waiting for a little while to be released, and we might release the recently added features, rather sooner than later.
Is #385 "ready" for most of the work? If yes, why still in draft? You might undraft it, so we can start reviewing more in depth, and working on it.
(I know it's off topic to talk about release process in this PR, but would like to talk about it here, as it's easier)

You can rebase if you want, but there are no conflicts and the difference with `main` is very little, it should have no impacts  with this PR.

**Comment by klaernie** [comment]
> Also, do you think, we should wait #385 to release v3.1.0? I feel like there are already lot of changes waiting for a little while to be released, and we might release the recently added features, rather sooner than later. Is #385 "ready" for most of the work? If yes, why still in draft? You might undraft it, so we can start reviewing more in depth, and working on it. (I know it's off topic to talk about release process in this PR, but would like to talk about it here, as it's easier)

I think it would be easiest to stuff it all into a v3.2.0 - #385 and #371 kind of go hand in hand - #371 takes care of the Makefile-based builds, and #385 revamps the official release process to entire be driven by goreleaser, making loads of the Makefile irrelevant. But for both we first need to ensure the wrapper projects are aware of the name change. On the other hand having an automatic release process would be nice to iterate quicker. I actually forgot to set it out of Draft status, my "check the existing archives" was my target for it being reviewable, but I totally overlooked that.
(I'm gonna copy that over to #385)

Your way forward with this branch sounds about right - I think the biggest problem is that even a small team not always has three people available (one to code, two to review), so it can get lengthy sometimes. And this only quadruples, when it's 32 commits and 1000 line changes, simply increasing the mental "do I really want to tackle this big PR now?" ;) Well, I definitely learned from that again.

File: cmd/editorconfig-checker/main_test.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+func captureReturncode(code int) {
+	mainHasRun <- code
+	runtime.Goexit()
+}
+
+func setArguments(args ...string) {
+	// os.Args needs to have the binary
+	os.Args = append([]string{"editorconfig-checker"}, args...)
+}
```
Comment (ccoVeille) [inline cmd/editorconfig-checker/main_test.go:19]: This is not a good way to do.

I would suggest this


```suggestion
func setArguments(t *testing.T, args ...string) {
	t.Helper()

	initialArgs := os.Args

	// os.Args needs to have the binary
	os.Args = append([]string{"editorconfig-checker"}, args...)
	t.Cleanup(func() {
		os.Args = initialArgs
	})
}
```

File: cmd/editorconfig-checker/main.go [hunk 8]
```diff
@@ -2,56 +2,69 @@
...
 var currentConfig *config.Config
 
-// Init function, runs on start automagically
-func init() {
-	var configFilePath string
-	var tmpExclude string
-	var c config.Config
-	var init bool
+// exitStub is there to be replaced while running the tests
+var exitStub = os.Exit
```
Comment (ccoVeille) [inline cmd/editorconfig-checker/main.go:30]: I won't call it stub here. What you do in the tests is a replacing it with a stub.

The calls made in main func with "exitStub" are very strange to me

Here it's the name that will be called in the code.

I could puld use something like this


```suggestion
var exiter = os.Exit
```

```suggestion
var exitFunc = os.Exit
```

**Comment by theoludwig** [comment]
> You can rebase if you want, but there are no conflicts and the difference with main is very little, it should have no impacts with this PR.

This PR will now need to be rebased. I advise to wait #389 to be merged, so that you only resolve conflicts once.

File: cmd/editorconfig-checker/main_test.go (hunk 7 above)
Comment (klaernie) [reply cmd/editorconfig-checker/main_test.go:19]: Would you just call it bad style to not restore the original arguments, or do you have any practical concerns? The only other consumer of `os.Args` is the `testing` package itself, but it has completed it's `os.Args` parsing by then.

File: cmd/editorconfig-checker/main.go (hunk 8 above)
Comment (klaernie) [reply cmd/editorconfig-checker/main.go:30]: You're right, technically speaking this would make this function in main.go `exitProxy`.
I'm not really a fan of `exiter` since it's too close to the stem "exciting", and just appending `Func` is also not my cup of tea. I'd hope you can accept `exitProxy`.

File: cmd/editorconfig-checker/main_test.go (hunk 7 above)
Comment (klaernie) [reply cmd/editorconfig-checker/main_test.go]: I reworked it according to your suggestions

**Comment by klaernie** [comment]
I rebased on top of the current main branch, incorporating the changes requested so far.

File: pkg/config/config.go [hunk 9]
```diff
@@ -116,11 +116,14 @@ type DisabledChecks struct {
...
 func NewConfig(configPaths []string) (*Config, error) {
 	if len(configPaths) == 0 {
-		return nil, fmt.Errorf("No config paths provided")
+		return nil, fmt.Errorf("no config paths provided")
 	}
 
 	var config Config
 
+	// initialize the Logger early, so usage of it does not fail
+	config.Logger.Init()
```
Comment (ccoVeille) [inline pkg/config/config.go:125]: I would have added a Init to config, that would have called the Logger Init

This way, if the logger changes nothing needs to be updated outside Config

But why not, it's the NewConfig constructor

File: pkg/config/config.go [hunk 10]
```diff
@@ -278,7 +284,7 @@ func (c Config) GetExcludesAsRegularExpression() string {
 // Save saves the config to it's Path
 func (c Config) Save(version string) error {
 	if utils.IsRegularFile(c.Path) {
-		return fmt.Errorf("File `%v` already exists", c.Path)
+		return fmt.Errorf("file `%v` already exists", c.Path)
```
Comment (ccoVeille) [inline pkg/config/config.go:287]: There is no reason to use %v for a string

So, this would be preferred

```suggestion
		return fmt.Errorf("file `%s` already exists", c.Path)
```

Also, this could be considered


```suggestion
		return fmt.Errorf("file %q already exists", c.Path)
```

File: cmd/editorconfig-checker/main.go [hunk 11]
```diff
@@ -66,88 +79,90 @@ func init() {
 	currentConfig, _ = config.NewConfig(configPaths)
```
[the remark this answers is not in the document]
Comment (klaernie) [reply cmd/editorconfig-checker/main.go:79]: it's not needed for `.NewConfig`, but for `.ParseConfig` - where it is present

File: cmd/editorconfig-checker/main.go [hunk 12]
```diff
@@ -2,57 +2,70 @@
...
 	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/outputformat"
 	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/utils"
 	"github.com/editorconfig-checker/editorconfig-checker/v3/pkg/validation"
 )
 
-// version is used vor the help
+// version is used for the help and to verify against the version stored in the config file
 // version is dynamically set at compiletime
-var version string
+var version string = "v3.0.3"
```
[the remark this answers is not in the document]
Comment (klaernie) [reply cmd/editorconfig-checker/main.go:22]: I would leave coming up with a better strategy up to #385 - here I just rebased my change to work nicely with #385 ( I introduced a magic string "development" as the version content and checked for that in a piece later, but this will no longer be needed after #385, maybe in two days time)

File: pkg/config/config.go (hunk 9 above)
Comment (klaernie) [reply pkg/config/config.go]: I later switched to a lazy initialization, which is more reliable, so this was not needed at all. I removed it.

File: cmd/editorconfig-checker/main_test.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+	lastSeenCode := <-mainHasRun
+	if lastSeenCode != 0 {
+		t.Errorf("main exited with return code %d, but we expected 0", lastSeenCode)
+	}
+}
+
+func TestMainInitializingANewConfig(t *testing.T) {
+	dir := t.TempDir()
+	setArguments(t, "--debug", "--verbose", "--init", "--config", dir+"/testwriteconfig.json")
+	go main()
```
[the remark this answers is not in the document]
Comment (klaernie) [reply cmd/editorconfig-checker/main_test.go]: good point, done.

File: pkg/config/config.go (hunk 10 above)
Comment (klaernie) [reply pkg/config/config.go]: I decided to keep the quotes - if for some obscure reason `c.Path` is empty or only filled with unprintable characters one would no longer recognize that this error message was supposed to contain a path.
But I replaced the default rendering with string rendering.

File: cmd/editorconfig-checker/main_test.go [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+		os.Args = initialArgs
+	})
+
+	// os.Args needs to have the binary as the first element
+	os.Args = append([]string{"editorconfig-checker"}, args...)
+}
+
+func TestMainOurCodebase(t *testing.T) {
+	cdRelativeToRepo(t, "")
+	setArguments(t, "--debug", "--verbose", "--exclude", "\\.git", "--exclude", "\\.exe$")
```
[the remark this answers is not in the document]
Comment (klaernie) [reply cmd/editorconfig-checker/main_test.go]: Done in all four strings.

File: cmd/editorconfig-checker/main.go [hunk 15]
```diff
@@ -66,88 +79,90 @@ func init() {
...
 	config.Logger.Verbose("%d files checked", len(filePaths))
 
-	if errorCount != 0 {
-		os.Exit(1)
+	if error.GetErrorCount(errors) != 0 {
+		exitProxy(1)
 	}
 
-	os.Exit(0)
+	exitProxy(0)
```
[the remark this answers is not in the document]
Comment (klaernie) [reply cmd/editorconfig-checker/main.go]: Done - including a test case that ensures nobody ever touches the values of these constants.

File: cmd/editorconfig-checker/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (klaernie) [reply cmd/editorconfig-checker/main.go]: I actually now took a different route - it turns out `.NewConfig` does not need to produce errors at all, so I removed the error from it's return values.

**Comment by klaernie** [comment]
I again incorporated all the feedback, I think this is the time for the final review now, unless something unexpected turns up

File: cmd/editorconfig-checker/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ccoVeille) [reply cmd/editorconfig-checker/main.go]: Thanks

File: cmd/editorconfig-checker/main_test.go [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+	"bytes"
+	"os"
+	"runtime"
+	"strings"
+	"testing"
+)
+
+var mainHasRun chan int
+
+func captureReturncode(code int) {
```
Comment (ccoVeille) [inline cmd/editorconfig-checker/main_test.go:13]:
```suggestion
func captureReturnCode(code int) {
```

File: cmd/editorconfig-checker/main_test.go [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+	}
+}
+
+func TestMainLoadingAncientConfig(t *testing.T) {
+	output, lastSeenCode := runWithArguments(t, "--debug", "--verbose", "--config", "testdata/ancient-config.json")
+	if lastSeenCode != exitCodeErrorOccurred {
+		t.Errorf("main exited with return code %d, but we expected %d", lastSeenCode, exitCodeErrorOccurred)
+	}
+	if !strings.Contains(output, "SpacesAftertabs") {
+		t.Errorf("main did not produce a warning that SpacesAftertabs is deprecated\nOutput:\n%s", output)
```
Comment (ccoVeille) [inline cmd/editorconfig-checker/main_test.go:87]: I'm surprised it's SpaceAftertabs and not SpaceAfterTabs

But I assume it's out of the PR scope

**Review by ccoVeille — APPROVED** [review]

File: cmd/editorconfig-checker/main_test.go (hunk 17 above)
Comment (klaernie) [reply cmd/editorconfig-checker/main_test.go:87]: we're testing that  the deprecation warning is output, not the presence of the new option.

**Review by theoludwig — APPROVED** [review]

File: cmd/editorconfig-checker/main_test.go (hunk 16 above)
Comment (klaernie) [reply cmd/editorconfig-checker/main_test.go]: hm.. I never knew I treated "returncode" as a single word.. although I guess proper english would write that as "return code"... Anyway, fixed it.

**Comment by klaernie** [comment]
I think if nothing else comes up, we should start the 72h now ;)
